partiql-planner/src/main/kotlin/org/partiql/planner/internal/window/WindowFunctionSignatureProvider.kt

```diff
@@ -0,0 +1,56 @@
+package org.partiql.planner.internal.window
+
+import org.partiql.plan.WindowFunctionSignature
+import org.partiql.planner.internal.ir.Rex
+import org.partiql.planner.internal.typer.DynamicTyper
+import org.partiql.spi.types.PType
+
+/**
+ * Provides the [WindowFunctionSignature] for a given window function name and arguments.
+ */
+internal object WindowFunctionSignatureProvider {
+
+    /**
+     * Returns the [WindowFunctionSignature] for the given window function name and arguments.
+     * For now, we only support RANK (0 args), DENSE_RANK (0 args), ROW_NUMBER (0 args), LAG (3 args), and LEAD (3 args).
+     * @param name the name of the window function.
+     * @param args the arguments to the window function.
+     * @return null if unable to create a signature.
+     */
+    fun get(name: String, args: List<Rex>, ignoreNulls: Boolean): WindowFunctionSignature? {
+        return when (args.size) {
+            0 -> when (name.lowercase()) {
+                "rank" -> RANK
+                "dense_rank" -> DENSE_RANK
+                "row_number" -> ROW_NUMBER
+                else -> null
+            }
+            3 -> when (val n = name.lowercase()) {
+                "lag", "lead" -> lagOrLead(n, args[0], args[1], args[2], ignoreNulls)
+                else -> null
+            }
+            else -> null
+        }
+    }
+
+    private val RANK = WindowFunctionSignature("rank", emptyList(), PType.bigint(), false)
+    private val DENSE_RANK = WindowFunctionSignature("dense_rank", emptyList(), PType.bigint(), false)
+    private val ROW_NUMBER = WindowFunctionSignature("row_number", emptyList(), PType.bigint(), false)
+
+    /**
+     * Creates the LAG window function signature.
+     * TODO: We may eventually want/need to handle coercion of the expr/default. This would require this class
+     *  to return a nullable [org.partiql.plan.WindowFunctionNode], not a [WindowFunctionSignature].
+     * @return null if unable to create a signature.
+     */
+    private fun lagOrLead(name: String, expr: Rex, offset: Rex, default: Rex, ignoreNulls: Boolean): WindowFunctionSignature? {
+        if (offset.type.code() != PType.BIGINT) {
+            return null
+        }
+        val returnTyper = DynamicTyper()
+        returnTyper.accumulate(expr)
+        returnTyper.accumulate(default)
```

Comment:
Do we want to accumulate or throw on mismatch?

Reply:
I might be missing something. The lag/lead's output type will be the expr or default if the offset row doesn't exist, so they don't need to match. In the case of coercible types, maybe we coerce them to their common super type, hence the TODO. But, I don't see any problem with the output type being ANY if the types aren't coercible.

The only validation I'd say is required is that the offset by coercible to BIGINT (or for now just BIGINT).